source/extensions/quic_listeners/quiche/platform/quic_server_stats_impl.h

```diff
@@ -6,18 +6,184 @@
+
+namespace {
+
+// In order to detect deprecated stats, these entries need to be declared in a struct as public
+// fields with corresponding getters. And they should be only accessed via their getters as if they
+// are private field. The reason for them to be public is for the accessibility of stats like
+// "a.b.c.d". When a stats is removed from QUICHE code, its getter is left unused. And this struct
+// must be declared in an annonymous namespace. This struct looks like:
+//  struct QuicheStats {
+//    // QUIC_SERVER_HISTOGRAM_BOOL(a.b.c.d, ...)
+//    // QUIC_SERVER_HISTOGRAM_ENUM(a.b.c.e, ...)
+//    struct a_t{
+//      struct b_t{
+//        struct c_t{
+//          Envoy::Stats::Metric& d;
+//          Envoy::Stats::Metric& d_() { return d; }
+//          EnumMap e;
+//          EnumMap& e_() { return e; }
+//        } c;
+//      } b;
+//    } a;
+//
+//    // QUIC_SERVER_HISTOGRAM_COUNTER(aaa, ...)
+//    Envoy::Stats::Histogram& aaa;
+//    Envoy::Stats::Histogram& aaa_() { return aaa; }
+//
+//    // QUIC_SERVER_HISTOGRAM_ENUM(bbb, ...)
+//    // Counters are not instantiated until incremented.
+//    EnumMap bbb;
+//    EnumMap& bbb_() { return bbb; }
+//    ...
+//  };
+struct QuicheStats {
+  QUICHE_STATS(STRUCT_MEMBER_COUNTER, STRUCT_MEMBER_GAUGE, STRUCT_MEMBER_HISTOGRAM,
+               STRUCT_MEMBER_ENUM_HISTOGRAM, STRUCT_MEMBER_STRUCT)
+};
+
+} // namespace
+
+class QuicStatsImpl {
+public:
+  void initializeStats(Envoy::Stats::Scope& scope) {
+    scope_ = &scope;
+    std::string prefix("quiche");
+    stats_.reset(
+        new QuicheStats{// // QUIC_SERVER_HISTOGRAM_COUNTER(aaa, ...)
+                        // scope.histogram("quiche.aaa", ...),
+                        //
+                        // // QUIC_SERVER_HISTOGRAM_ENUM(bbb, ...)
+                        // {},
+                        //
+                        // // QUIC_SERVER_HISTOGRAM_ENUM(a.b.c.d, ...)
+                        // // QUIC_SERVER_HISTOGRAM_ENUM(a.b.c.e, ...)
+                        // {{{ {},{} }}},
+                        QUICHE_STATS(INIT_STATS_COUNTER, INIT_STATS_GAUGE, INIT_STATS_HISTOGRAM,
+                                     INIT_STATS_ENUM_HISTOGRAM, INIT_STATS_STRUCT)});
+  }
+
+  void resetForTest() {
+    stats_.reset();
+    scope_ = nullptr;
+  }
+
+  Envoy::Stats::Counter& createCounter(std::string name) { return scope_->counterFromString(name); }
+
+  QuicheStats& stats() {
+    RELEASE_ASSERT(stats_ != nullptr, "Quiche stats is not initialized_.");
+    return *stats_;
+  }
+
+private:
+  Envoy::Stats::Scope* scope_;
+  std::unique_ptr<QuicheStats> stats_;
+};
+
+QuicStatsImpl& getQuicheStats() {
+  // Each worker thread has its own set of references to counters, gauges and
+  // histograms.
+  static thread_local QuicStatsImpl stats_impl = QuicStatsImpl();
+  return stats_impl;
+}
+
+#define QUIC_SERVER_HISTOGRAM_ENUM_IMPL(enum_name, sample, enum_size, docstring)                   \
```

Comment:
can we please make this a wrapper to an actual function-call where you do all the stringification you need in the macro and do the real work in the function? I really don't like these huge macros. Hard to debug, slow to compile.